File: python/lld/board-games/snake-ladder/engine.py
import random

from snake_ladder_config import SnakeLadderConfig, DEFAULT_CONFIG

TOKEN_PIECES = [
    Piece('\U0001f534', '#EF4444'),
    Piece('\U0001f535', '#3B82F6'),
    Piece('\U0001f7e2', '#10B981'),
    Piece('\U0001f7e1', '#F59E0B'),
]

DEFAULT_SL_PLAYERS = [
    Player('Player 1', TOKEN_PIECES[0]),
    Player('Player 2', TOKEN_PIECES[1]),
]

class SnakeLadderEngine:
    def __init__(self, players: List[Player] = None, config: SnakeLadderConfig = None):
        self.players = players or DEFAULT_SL_PLAYERS
        self.config = config or DEFAULT_CONFIG
        self.positions: Dict[str, int] = {}
        self.current_player_index = 0
        self.status = GameStatus.NOT_STARTED
        self.move_history: List[SnakeLadderMove] = []
        self.winner: Optional[Player] = None

    def start(self):
        self.status = GameStatus.IN_PROGRESS
        self.current_player_index = 0
        self.move_history = []
        self.winner = None
        self.positions = {p.name: 0 for p in self.players}

    def roll_dice(self) -> int:
        return random.randint(1, 6)

    def make_move(self, player: Player, dice_value: int) -> Dict:
        if self.status != GameStatus.IN_PROGRESS:
            return {'success': False, 'error': 'Game not in progress', 'move': None}

        current_pos = self.positions.get(player.name, 0)
        new_pos = current_pos + dice_value

        if new_pos > self.config.finish:
            new_pos = current_pos
            move = SnakeLadderMove(player, dice_value, current_pos, new_pos)
            self.move_history.append(move)
            self.current_player_index = (self.current_player_index + 1) % len(self.players)
            return {'success': True, 'error': None, 'move': move}

        landed_on_snake = False
        landed_on_ladder = False

        if new_pos in self.config.snakes:
            new_pos = self.config.snakes[new_pos]
            landed_on_snake = True
        elif new_pos in self.config.ladders:
            new_pos = self.config.ladders[new_pos]
            landed_on_ladder = True

        self.positions[player.name] = new_pos
        move = SnakeLadderMove(player, dice_value, current_pos, new_pos,
                                landed_on_snake, landed_on_ladder)
        self.move_history.append(move)

        if new_pos == self.config.finish:
            self.status = GameStatus.COMPLETED
            self.winner = player
        else:
            self.current_player_index = (self.current_player_index + 1) % len(self.players)

        return {'success': True, 'error': None, 'move': move}

    def get_position(self, player: Player) -> int:
        return self.positions.get(player.name, 0)

    def get_current_player(self) -> Player:
        return self.players[self.current_player_index]

    def get_status(self) -> GameStatus:
        return self.status

    def get_winner(self) -> Optional[Player]:
        return self.winner

    def get_board_cells(self) -> List[Dict]:
        cells = []
        for i in range(1, self.config.finish + 1):
            occupying = [p for p in self.players if self.positions.get(p.name) == i]
            entry = {'cell': i, 'players': occupying}
            if i in self.config.snakes:
                entry['snake'] = self.config.snakes[i]
            if i in self.config.ladders:
                entry['ladder'] = self.config.ladders[i]
            cells.append(entry)
        return cells

File: go/lld/board-games/snake-ladder/engine.go
package game

import "math/rand"

var tokenPieces = []Piece{
    {Symbol: "🔴", Color: "#EF4444"},
    {Symbol: "🔵", Color: "#3B82F6"},
    {Symbol: "🟢", Color: "#10B981"},
    {Symbol: "🟡", Color: "#F59E0B"},
}

type SnakeLadderEngine struct {
    players            []Player
    config             SnakeLadderConfig
    positions          map[string]int
    currentPlayerIndex int
    status             GameStatus
    moveHistory        []SnakeLadderMove
    winner             *Player
}

func NewSnakeLadderEngine(players []Player) *SnakeLadderEngine {
    return NewSnakeLadderEngineWithConfig(players, DefaultConfig())
}

func NewSnakeLadderEngineWithConfig(players []Player, config SnakeLadderConfig) *SnakeLadderEngine {
    if players == nil {
        players = []Player{
            {Name: "Player 1", Piece: tokenPieces[0]},
            {Name: "Player 2", Piece: tokenPieces[1]},
        }
    }
    e := &SnakeLadderEngine{players: players, config: config}
    e.Start()
    return e
}

func (e *SnakeLadderEngine) Start() {
    e.status = InProgress
    e.currentPlayerIndex = 0
    e.moveHistory = nil
    e.winner = nil
    e.positions = make(map[string]int)
    for _, p := range e.players {
        e.positions[p.Name] = 0
    }
}

func (e *SnakeLadderEngine) RollDice() int {
    return rand.Intn(6) + 1
}

type MoveResult struct {
    Success bool
    Error   string
    Move    *SnakeLadderMove
}

func (e *SnakeLadderEngine) MakeMove(player Player, diceValue int) MoveResult {
    if e.status != InProgress {
        return MoveResult{Success: false, Error: "Game not in progress"}
    }

    currentPos := e.positions[player.Name]
    newPos := currentPos + diceValue

    if newPos > e.config.Finish {
        newPos = currentPos
        m := &SnakeLadderMove{Player: player, DiceValue: diceValue, FromCell: currentPos, ToCell: newPos}
        e.moveHistory = append(e.moveHistory, *m)
        e.currentPlayerIndex = (e.currentPlayerIndex + 1) % len(e.players)
        return MoveResult{Success: true, Move: m}
    }

    landedOnSnake := false
    landedOnLadder := false

    if tail, ok := e.config.Snakes[newPos]; ok {
        newPos = tail
        landedOnSnake = true
    } else if top, ok := e.config.Ladders[newPos]; ok {
        newPos = top
        landedOnLadder = true
    }

    e.positions[player.Name] = newPos
    m := &SnakeLadderMove{
        Player: player, DiceValue: diceValue,
        FromCell: currentPos, ToCell: newPos,
        LandedOnSnake: landedOnSnake, LandedOnLadder: landedOnLadder,
    }
    e.moveHistory = append(e.moveHistory, *m)

    if newPos == e.config.Finish {
        e.status = Completed
        e.winner = &player
    } else {
        e.currentPlayerIndex = (e.currentPlayerIndex + 1) % len(e.players)
    }

    return MoveResult{Success: true, Move: m}
}

func (e *SnakeLadderEngine) GetPosition(player Player) int {
    return e.positions[player.Name]
}

func (e *SnakeLadderEngine) GetCurrentPlayer() Player {
    return e.players[e.currentPlayerIndex]
}

func (e *SnakeLadderEngine) GetStatus() GameStatus {
    return e.status
}

func (e *SnakeLadderEngine) GetWinner() *Player {
    return e.winner
}

type BoardCell struct {
    Cell    int
    Players []Player
    Snake   int
    Ladder  int
    HasSnake  bool
    HasLadder bool
}

func (e *SnakeLadderEngine) GetBoardCells() []BoardCell {
    cells := make([]BoardCell, 0, e.config.Finish)
    for i := 1; i <= e.config.Finish; i++ {
        var occupying []Player
        for _, p := range e.players {
            if e.positions[p.Name] == i {
                occupying = append(occupying, p)
            }
        }
        entry := BoardCell{Cell: i, Players: occupying}
        if tail, ok := e.config.Snakes[i]; ok {
            entry.Snake = tail
            entry.HasSnake = true
        }
        if top, ok := e.config.Ladders[i]; ok {
            entry.Ladder = top
            entry.HasLadder = true
        }
        cells = append(cells, entry)
    }
    return cells
}